### PR TITLE
api: add docs for live_update

### PR DESCRIPTION
**Pull request by landism**
This is just api docs to ensure we're on the same page. I've filed ch2047 for friendlier non-api docs.

This won't be merged until after the implementation is in master.

### CODE REVIEW

File: api/api.py [hunk 1]
```diff
@@ -294,3 +294,64 @@ def custom_build(ref: str, command: str, deps: List[str], tag: str = "", disable
...
   pass
+
+class LiveUpdateStep:
+    """A step in the process of performing a LiveUpdate on an image's container.
+
+    See :meth:`live_update`.
+    """
+    pass
+
+def workdir(dir: str) -> LiveUpdateStep:
```
Comment (maiamcc) [inline api/api.py]: also sync "dest"'s, if relative, are evaluated relative to this. maybe worth a "see also" to docker's WORKDIR command?

Comment (maiamcc) [reply api/api.py]: specify that we only allow one and it must come first

File: api/api.py [hunk 2]
```diff
@@ -294,3 +294,64 @@ def custom_build(ref: str, command: str, deps: List[str], tag: str = "", disable
...
+
+    - 0 or 1 :meth:`workdir`
+    - 0 or more :meth:`sync`
+    - 0 or more :meth:`run`
+    - 0 or 1 :meth:`container_restart`
+
+    When a file changes:
+
+    1. If it matches any of the paths in `full_rebuild_trigger`, a full rebuild + deploy will be executed.
+    2. Otherwise, if it matches any of the local paths in `sync` steps, a live update will be executed.
```
Comment (maiamcc) [inline api/api.py]: ...a live update will be executed:
- copy any changed sync'd files to the container
- execute relevant run's
- if container_restart provided, restart the container

(or something)

File: api/api.py (hunk 1 above)
Comment (maiamcc) [reply api/api.py]: oh hmm you list this below... idk, maybe worth listing here as well? (and similarly under container_restart?) 🤷‍♀️

File: api/api.py [hunk 3]
```diff
@@ -294,3 +294,65 @@ def custom_build(ref: str, command: str, deps: List[str], tag: str = "", disable
...
+    pass
+
+def run(cmd: str, trigger: Union[List[str], str] = []) -> LiveUpdateStep:
+    """Specify that the given `cmd` should be executed when updating an image's container
+
+    May not precede any `sync` steps in a `live_update`.
+
+    Args:
+      cmd: A shell command.
+      trigger: If the ``trigger`` argument is specified, the build step is only run on changes to the given file(s).
```
Comment (jazzdan) [inline api/api.py]:
```suggestion
      trigger: If the ``trigger`` argument is specified, the build step is only run on changes to the given file(s). Paths relative to Tiltfile.
```

File: api/api.py [hunk 4]
```diff
@@ -294,3 +294,65 @@ def custom_build(ref: str, command: str, deps: List[str], tag: str = "", disable
...
+
+    """
+    pass
+
+def live_update(image: str, steps: List[LiveUpdateStep], full_rebuild_trigger: Union[str, List[str]] = []) -> None:
+    """Specify how to update containers running `image`.
+
+    The list of steps must be, in order:
+
+    - 0 or more :meth:`sync`
```
Comment (nicks) [inline api/api.py]: meth? :eyes:

Comment (landism) [reply api/api.py]: this makes `sync` a hyperlink to the method `sync`, instead of just monospaced text

https://stackoverflow.com/a/22714510

Comment (nicks) [reply api/api.py]: :breaking-bad-this-is-not-meth-dot-gif:

**Comment by maiamcc** [comment]
I don't need a full re-review, but but appreciate a quick skim in re: layout and organization

**Comment by maiamcc** [comment]
cc @landism i can't request you as a reviewer of your own PR but hiii

File: api/api.py [hunk 5]
```diff
@@ -7,7 +7,60 @@ class Blob:
 
    To wrap a string as a blob, call ``blob(my_str)``"""
 
-def docker_build(ref: str, context: str, build_args: Dict[str, str] = {}, dockerfile: str = "Dockerfile", dockerfile_contents: Union[str, Blob] = "") -> None:
+class LiveUpdateStep:
+  """A step in the process of performing a LiveUpdate on an image's container.
```
Comment (landism) [inline api/api.py]: Is it worth linking all the LiveUpdateStep constructors from here, or let the docs handle that?

File: api/api.py [hunk 6]
```diff
@@ -7,7 +7,60 @@ class Blob:
...
+  pass
+
+def run(cmd: str, trigger: Union[List[str], str] = []) -> LiveUpdateStep:
+  """Specify that the given `cmd` should be executed when updating an image's container
+
+  May not precede any `sync` steps in a `live_update`.
+
+  Args:
+    cmd: A shell command.
+    trigger: If the ``trigger`` argument is specified, the build step is only run on changes to the given file(s). Paths relative to Tiltfile.
```
Comment (landism) [inline api/api.py]:
```suggestion
    trigger: If the ``trigger`` argument is specified, the build step is only run when there are changes to the given file(s). Paths relative to Tiltfile.
```

(my first read of the old wording was like "oh it'll somehow intelligently only run go build on the files that changed - wait how can it do that?")

**Comment by landism** [comment]
LGTM

File: api/api.py (hunk 5 above)
Comment (maiamcc) [reply api/api.py]: UHHHH :-/ i'm still holding out hope that we can get the nesting to work. let me try that once we get the stuff merged, if not i'll go back and link stuff

**Pull request merged**